File: services/learning/src/learning/adaptive/flags.py
"""Feature-flag client for Adaptive Engine. GAP-16 wire-up #4.

Consumes:
- `irt_model_enabled` (fallback FALSE) — when ON, the 3PL IRT cold-start path runs;
  when OFF, fall back to binary-search cold-start (Sprint 1 default; SPIKE-01 informs Sprint 2).
"""

from __future__ import annotations

import logging

from alp_flags import FlagClient, structlog_decision_hook

from learning.adaptive.config import settings

log = logging.getLogger(__name__)

FALLBACKS: dict[str, bool] = {
    "irt_model_enabled": False,
}

_client: FlagClient | None = None


def client() -> FlagClient:
    if _client is None:
        raise RuntimeError("FlagClient not initialised — connect_flags() must run in app lifespan")
    return _client


async def connect_flags() -> None:
    global _client
    _client = FlagClient(
        on_decision=structlog_decision_hook("adaptive-engine"),
        institution_url=settings.institution_base_url,
        nats_url=settings.nats_url,
        fallbacks=FALLBACKS,
        cache_ttl=30.0,
    )
    await _client.connect()
    log.info("adaptive_engine flag client connected")


async def close_flags() -> None:
    global _client
    if _client is not None:
        await _client.close()
        _client = None


async def use_irt(tenant_id: str | None = None) -> bool:
    try:
        return await client().evaluate("irt_model_enabled", tenant_id=tenant_id)
    except Exception:  # noqa: BLE001
        return False

File: services/quiz/internal/flags/flags.go
// Package flags wires the alpflags SDK into the Quiz service.
//
// GAP-16: Quiz consumes:
//   - irt_model_enabled (default false) — when true, use 3PL IRT cold-start;
//     when false, use binary-search cold-start (Sprint 1 default; SPIKE-01 informs Sprint 2 flip).
package flags

import (
	"context"
	"log/slog"
	"os"
	"time"

	"github.com/adaptive-learn/alpflags"
)

// Fallbacks declares every flag Quiz consumes plus its hardcoded fallback.
// Missing fallback at evaluation time is a build-time bug.
var Fallbacks = map[string]bool{
	"irt_model_enabled": false,
}

// New constructs a flag client from environment configuration.
// Returns a client that has been Connect()'d (NATS subscription up, if reachable).
func New(ctx context.Context, logger *slog.Logger) (*alpflags.Client, error) {
	// Read both names — QUIZ_INSTITUTION_BASE_URL is the canonical one (matches
	// internal/config); QUIZ_INSTITUTION_URL was used by an earlier draft. Keep
	// both so existing dev .env files don't silently fall back to localhost.
	institutionURL := envDefault("QUIZ_INSTITUTION_BASE_URL", envDefault("QUIZ_INSTITUTION_URL", "http://localhost:38008"))
	natsURL := envDefault("QUIZ_NATS_URL", "nats://localhost:34222")

	c := alpflags.New(alpflags.Options{
		InstitutionURL: institutionURL,
		NatsURL:        natsURL,
		Fallbacks:      Fallbacks,
		CacheTTL:       30 * time.Second,
		Logger:         logger,
		OnDecision:     alpflags.SlogDecisionHook("quiz", logger),
	})
	if err := c.Connect(ctx); err != nil {
		return nil, err
	}
	return c, nil
}

func envDefault(key, def string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return def
}
